pyunlocbox/acceleration.py

```diff
@@ -0,0 +1,502 @@
+        the garbage collector can free the memory. Gets called when
+        :func:`pyunlocbox.solvers.solve` finishes running.
+        """
+        self._post()
+        del self.sol
+
+    def _post(self):
+        raise NotImplementedError("Class user should define this method.")
+
+
+class dummy(accel):
+    r"""
+    Dummy acceleration scheme.
+
+    Used by default in most of the solvers. It simply returns unaltered the
+    step size and solution point it receives.
+    """
+
+    def _pre(self, functions, x0):
+        pass
+
+    def _update_step(self, solver, objective, niter):
+        return solver.step
+
+    def _update_sol(self, solver, objective, niter):
+        # Track the solution, but otherwise do nothing
+        self.sol[:] = solver.sol
+        return solver.sol
+
+    def _post(self):
+        pass
+
+
+# -----------------------------------------------------------------------------
+# Stepsize optimizers
+# -----------------------------------------------------------------------------
+
+
+class backtracking(dummy):
+    r"""
+    Backtracking based on a local quadratic approximation of the objective.
+
+    Parameters
+    ----------
+    eta : float
+        A number between 0 and 1 representing the ratio of the geometric
+        sequence formed by successive step sizes. In other words, it
+        establishes the relation `step_new = eta * step_old`.
+        Default is 0.5.
+
+    Notes
+    -----
+    This is the backtracking strategy proposed in the original FISTA paper,
+    :cite:`beck2009FISTA`.
+
+    Examples
+    --------
+    >>> from pyunlocbox import functions, solvers, acceleration
+    >>> import numpy as np
+    >>> y = [4, 5, 6, 7]
+    >>> x0 = np.zeros(len(y))
+    >>> f1 = functions.norm_l2(y=y)
+    >>> f2 = functions.dummy()
+    >>> accel = acceleration.backtracking()
+    >>> solver = solvers.forward_backward(accel=accel, step=0.5)
+    >>> ret = solvers.solve([f1, f2], x0, solver, atol=1e-5)
+    Solution found after 12 iterations:
+        objective function f(sol) = 7.510185e-06
+        stopping criterion: ATOL
+    >>> ret['sol']
+    array([ 3.99902344,  4.9987793 ,  5.99853516,  6.99829102])
+
+    """
+
+    def __init__(self, eta=0.5, **kwargs):
+        if (eta > 1) or (eta <= 0):
+            raise ValueError("eta must be between 0 and 1.")
+        self.eta = eta
+        super(backtracking, self).__init__(**kwargs)
+
+    def _pre(self, functions, x0):
+        self.smooth_funs = []  # Smooth functions.
+        for f in functions:
+            if 'GRAD' in f.cap(x0):
+                self.smooth_funs.append(f)
+
+    def _update_step(self, solver, objective, niter):
+        """
+        Notes
+        -----
+        TODO: For now we're recomputing gradients in order to evaluate the
+        backtracking criterion. In the future, it might be interesting to
+        think of some design changes so that this function has access to the
+        gradients directly.
```

Comment:
Although it's always a trade-off with memory usage. In this snippet, however, I have to call `eval` because the `solver` function has not updated the objective list yet at the moment this function is called. I think I'll leave the implementation like this for now, since it's simpler, but keep in mind possible changes in the future.